### PR TITLE
KQL Rewrite V2

### PR DESCRIPTION
**Pull request by AsafMah**

### CODE REVIEW

File: kusto/test/etoe/etoe_test.go [hunk 1]
```diff
@@ -581,6 +580,161 @@ func TestQueries(t *testing.T) {
...
+	}{
+		{
+			desc: "New query method",
+			stmt: kql.NewStatementBuilder("table(tableName) | where vstr == txt"),
+			options: []kusto.QueryOption{kusto.QueryParameters(*kql.NewStatementQueryParameters().
+				AddString("tableName", "string", allDataTypesTable).
+				AddString("txt", "string", "asdf"))},
+			qcall: client.Query,
+			doer: func(row *table.Row, update interface{}) error {
+				if row.Replace {
```
Comment (AsafMah) [inline kusto/test/etoe/etoe_test.go:637]: Why do you print the rows, and not, test them?

File: kusto/test/etoe/etoe_test.go [hunk 2]
```diff
@@ -581,6 +580,161 @@ func TestQueries(t *testing.T) {
...
+					fmt.Println("---") // Replace flag indicates that the query result should be cleared and replaced with this row
```
Comment (AsafMah) [inline kusto/test/etoe/etoe_test.go:638]: Do a test with all of the data types, making sure they all pass correctly and have the correct results.

File: kusto/kusto.go [hunk 3]
```diff
@@ -286,6 +280,17 @@
 			return nil, errors.ES(op, errors.KClientArgs, "QueryValues in the the Stmt were incorrect: %s", err).SetNoRetry()
 		}
 	}
+	if query.SupportsParameters() {
+		if len(opt.requestProperties.QueryParameters.ToParameterCollection()) != 0 {
+			return nil, errors.ES(op, errors.KClientArgs, "Cannot use both Stmt and StatementQueryParameters. Please use statementBuilder instead of Stmt.").SetNoRetry()
```
Comment (AsafMah) [inline kusto/kusto.go:285]: "kusto.Stmt does not support the QueryParameters option. Construct your query using kql.Builder"

File: kusto/kusto.go [hunk 4]
```diff
@@ -286,6 +280,17 @@
...
+		}
+		params, err := query.GetParameters()
+		if err != nil {
+			return nil, errors.ES(op, errors.KClientArgs, "QueryValues in the the Stmt were incorrect: %s", err).SetNoRetry()
```
Comment (AsafMah) [inline kusto/kusto.go:289]: Parameter validation error: %s

File: kusto/test/etoe/etoe_test.go [hunk 5]
```diff
@@ -581,6 +580,161 @@ func TestQueries(t *testing.T) {
...
+		doer func(row *table.Row, update interface{}) error
+		// gotInit creates the variable that will be used by doer's update argument.
+		gotInit func() interface{}
+		// want is the data we want to receive from the query.
+		want interface{}
+		// should the test fail
+		failFlag bool
+	}{
+		{
+			desc: "New query method",
```
Comment (AsafMah) [inline kusto/test/etoe/etoe_test.go:630]: E2E tests for building complex queries with Statement Builder

File: kusto/queryopts.go [hunk 6]
```diff
@@ -18,6 +19,7 @@
 	Parameters      map[string]string
 	Application     string
 	User            string
+	QueryParameters kql.StatementQueryParameters `json:"-"`
```
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/queryopts.go:22]: WDYM?

File: kusto/kql/query_parameters_test.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+				AddDatabase("database").AddLiteral(".").
+				AddTable("table").AddLiteral(" | where ").
+				AddColumn("column").AddLiteral(" == ").
+				AddFunction("function").AddLiteral("() ;"),
+			NewStatementQueryParameters().
+				AddLiteral("database", "string", "foo_1").
+				AddLiteral("table", "string", "_bar").
+				AddLiteral("column", "string", "_baz").
+				AddLiteral("function", "string", "func_"),
+			"declare query_parameters(database:string, table:string, column:string, function:string);\ndatabase(\"database\").table | where column == function() ;"},
```
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters_test.go:29]: WDYM? i agree addLiteral is a mistake, but wont AddString work?

[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/query_parameters_test.go:29]: The parameters are typed, if you pass a parameter of type string with the value "a" to "table", it will be:
database(\"database\")."a"
Which will not work.

Try running the generated query.

File: kusto/queryopts.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/queryopts.go:22]: Add `json:"-"` to everything except Options and Parameters

File: kusto/test/etoe/etoe_test.go [hunk 8]
```diff
@@ -581,6 +580,161 @@ func TestQueries(t *testing.T) {
 	}
 }
 
+func TestStatment(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/test/etoe/etoe_test.go:583]: Didn't fix

File: kusto/kql/value.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+func (b *statementBuilder) AddTimespan(value time.Duration) Builder {
+	return b.addBase(newValue(value, types.Timespan))
+}
+
+func (b *statementBuilder) AddDecimal(value decimal.Decimal) Builder {
+	return b.addBase(newValue(value, types.Decimal))
+}
+
+func (q *StatementQueryParameters) AddBool(key string, paramType string, value bool) *StatementQueryParameters {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/value.go:96]: didn't fix

File: kusto/kql/value.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+func (v *kqlValue) Type() types.Column {
+	return v.kustoType
+}
+
+func (v *kqlValue) String() string {
+	value := v.value
+	switch v.kustoType {
+	case types.String:
+		return value.(string)
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/value.go]: didn't fix

File: kusto/conn.go [hunk 11]
```diff
@@ -153,10 +153,17 @@
 
 	switch execType {
 	case execQuery, execMgmt:
+		var err error
+		var csl string
+		if query.SupportsParameters() {
+			csl = query.String()
+		} else {
+			csl = fmt.Sprintf("%s\n%s", properties.QueryParameters.ToDeclarationString(), query.String())
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/conn.go:161]: didn't fix

File: kusto/kql/query_parameters_test.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+func TestQueryParameters(t *testing.T) {
+	tests := []struct {
+		name     string
+		b        Builder
+		qp       *StatementQueryParameters
+		expected string
+	}{
+		{"Test empty", NewStatementBuilder(""), NewStatementQueryParameters(), "\n"},
+		{"Test add string", NewStatementBuilder(""), NewStatementQueryParameters().AddString("foo", "string", "bar"), "declare query_parameters(foo:string);\n"},
+		{"Test add identifiers", // test might fail at times due to the pseudo-random nature of map that will sometimes change the order of the declaration string.
```
Comment (AsafMah) [inline kusto/kql/query_parameters_test.go:18]: No, don't make tests that might fail. Have the check be more clever.

File: kusto/test/etoe/etoe_test.go [hunk 13]
```diff
@@ -581,6 +580,593 @@ func TestQueries(t *testing.T) {
...
+		doer func(row *table.Row, update interface{}) error
+		// gotInit creates the variable that will be used by doer's update argument.
+		gotInit func() interface{}
+		// want is the data we want to receive from the query.
+		want interface{}
+		// should the test fail
+		failFlag bool
+	}{
+		{
+			desc: "Complex query with Statement Builder - with String parameters",
```
Comment (AsafMah) [inline kusto/test/etoe/etoe_test.go:632]: Nice, but you don't need to overload the server with 10 tests. Make one test with all of the checks (with ands).

File: kusto/test/etoe/etoe_test.go [hunk 14]
```diff
@@ -581,6 +580,593 @@ func TestQueries(t *testing.T) {
...
+			stmt: kql.NewStatementBuilder("table(tableName) | where vstr == txt"),
```
Comment (AsafMah) [inline kusto/test/etoe/etoe_test.go:633]: Make an equivalent e2e test for the StatementBuilder

File: kusto/kql/query_parameters_test.go [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+func TestQueryParameters(t *testing.T) {
+	tests := []struct {
+		name     string
+		b        Builder
+		qp       *StatementQueryParameters
+		expected string
+	}{
+		{"Test empty", NewStatementBuilder(""), NewStatementQueryParameters(), "\n"},
+		{"Test add literal", NewStatementBuilder(""), NewStatementQueryParameters().AddLiteral("foo", "string", "bar"), "declare query_parameters(foo:string);\n"},
+		{"Test add identifiers", // test might fail at times due to the pseudo-random nature of map that will sometimes change the order of the declaration string.
```
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters_test.go:18]: what else should i test? (given im testing always in e2e)

**Review by AsafMah — COMMENTED** [review]
more comments

File: kusto/kql/query_parameters_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/query_parameters_test.go:18]: E2E tests are expansive, you can basically test just the functionality that queries and parameters work.
You want to test more combinations here - complicated parameters, escaping, etc

File: kusto/mock.go [hunk 16]
```diff
@@ -121,15 +121,15 @@
 type mockConn struct {
 }
 
-func (m mockConn) queryToJson(ctx context.Context, db string, query Stmt, options *queryOptions) (string, error) {
+func (m mockConn) queryToJson(ctx context.Context, db string, query Statement, options *queryOptions) (string, error) {
 	return "[]]", nil
```
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/mock.go:125]: if you're regarding the change to Statement, if i recall correctly one of the tests used a custom type that used it, so had to be changed. anyway, it doesn't change it's purpose as Stmt "inherits Statement

File: kusto/kql/query_parameters_test.go (hunk 12 above)
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters_test.go:18]: @yogilad you said it's not worth the time and just leave a note. what do we think we should do?

File: kusto/kql/query_parameters_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters_test.go:18]: So given that im testing pretty much that (all parameters, and escaping problem) in e2e, should i move it here? and what should be on e2e instead?

File: kusto/kql/value.go [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+func (b *statementBuilder) AddDecimal(value value.Decimal) Builder {
+	return b.addBase(newValue(value, types.Decimal))
+}
+
+func (q *StatementQueryParameters) AddBool(key string, value bool) *StatementQueryParameters {
+	return q.addBase(key, string(types.Bool), newValue(value, types.Bool))
+}
+
+func (q *StatementQueryParameters) AddDateTime(key string, value time.Time) *StatementQueryParameters {
```
Comment (AsafMah) [inline kusto/kql/value.go:99]: You're repeating yourself with string(types..) - Have addbase accept a value, and do the calc there.

File: kusto/kql/value.go [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func newValue(value interface{}, kustoType types.Column) Value {
+	return &kqlValue{
+		value:     value,
+		kustoType: kustoType,
+	}
+}
+
+func (b *statementBuilder) AddBool(value bool) Builder {
```
Comment (AsafMah) [inline kusto/kql/value.go:55]: Yochai thinks it's better if all of the methods from here down would be in the file with the struct except here.
Just cut + paste to builder and query_parameters

File: kusto/kql/identifier.go [hunk 19]
```diff
@@ -0,0 +1,45 @@
+package kql
+
+import "fmt"
+
+type Identifier struct {
```
Comment (AsafMah) [inline kusto/kql/identifier.go:5]: Yochai doesn't think this should be a struct.
Simple:
delete the struct.
For add database, just do `return b.addBase(fmt.Sprintf("%s(%s)", "database", QuoteString(database, false)))`

For the rest:
make normalize name into a function accepting name (we don't need the wrapping anymore), and change all to b.addBase(table) etc

File: kusto/kql/query_parameters_test.go (hunk 12 above)
Comment (AsafMah) [reply kusto/kql/query_parameters_test.go:18]: Spoke to Yochai - he said that the test shouldn't be flakey. Just check the substring.

File: kusto/kql/query_parameters_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/query_parameters_test.go:18]: I meant giving legitimate but weird names that need to be escaped. See builder_test for comparison, and also test the result of ToParameterCollection

File: kusto/kql/value.go [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+func (q *StatementQueryParameters) AddString(key string, value string) *StatementQueryParameters {
+	return q.addBase(key, string(types.String), newValue(value, types.String))
+}
+
+func (q *StatementQueryParameters) AddTimespan(key string, value time.Duration) *StatementQueryParameters {
+	return q.addBase(key, string(types.Timespan), newValue(value, types.Timespan))
+}
+
+func (q *StatementQueryParameters) AddDecimal(key string, value value.Decimal) *StatementQueryParameters {
```
Comment (AsafMah) [inline kusto/kql/value.go:131]: Why did you change it from decimal.decimal? now creating the value is much more awkward

File: kusto/kql/value.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/value.go]: So quoting it worked in parameters the entire time. Great!

File: kusto/conn.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/conn.go:161]: This isn't what I meant. Stmt is already safe. I meant that properties.QueryParameters.ToDeclarationString() might be a problem, but it seems to be getting a default value so it's ok. You can revert this change.

File: kusto/mock.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/mock.go:125]: just change it to "[]" from "[]]"

**Review by AsafMah — COMMENTED** [review]
More comments

File: kusto/conn.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/conn.go:161]: i dont think its a bad check anyway, isnt it? to make sure it indeed has parameters. WDYT?

File: kusto/kql/identifier.go (hunk 19 above)
Comment (ronmonetaMicro) [reply kusto/kql/identifier.go:5]: i do need though something that implements String (as addBase expects fmt.Stringer).  should i deliver the values the same way as i do in AddString perhaps (with the normalize)?  
b.addBase(newValue(NormalizeName(value), types.String))

File: kusto/kql/value.go (hunk 20 above)
Comment (ronmonetaMicro) [reply kusto/kql/value.go:131]: fitted it by a similar test. i can change it back i believe, but how is it less troublesome? ill still need 
kusto.ParamType{Type: types.Decimal, Default: "2.00000000000001"}
or something like that, no?

File: kusto/conn.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/conn.go:161]: They already check it in String()
You just add extra work of doing an additional GetParameters

File: kusto/kql/value.go (hunk 20 above)
Comment (AsafMah) [reply kusto/kql/value.go:131]: decimal.Decimal is created by decimal.RequireFromString/decimal.NewFromString, which are much simpler, and expecting the user to past {Valid: true} is weird.
types.Decimal is for our internal conversion from kusto.

File: kusto/conn.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (yogilad) [reply kusto/conn.go:161]: The condition below 
		if query.SupportsParameters() && query.HasParameters() {
Ought to be 
		if query is old format || new query params is empty
		
This makes sure new queries without params are not prepanded with newlines for no reason

File: kusto/kql/query_parameters.go [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+type ParamVals struct {
+	paramType string
+	value     string
+}
+
+type StatementQueryParameters struct {
+	parameters map[string]ParamVals
+}
+
+var buildPool = sync.Pool{
```
Comment (yogilad) [inline kusto/kql/query_parameters.go:18]: @AsafMah , do we want to keep a pool of string builders (memory management wise and such)?

File: kusto/kql/query_parameters.go [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+	New: func() interface{} {
+		return new(strings.Builder)
+	},
+}
+
+func NewStatementQueryParameters() *StatementQueryParameters {
+	return &StatementQueryParameters{parameters: make(map[string]ParamVals)}
+}
+func (q *StatementQueryParameters) addBase(key string, paramType string, value fmt.Stringer) *StatementQueryParameters {
```
Comment (yogilad) [inline kusto/kql/query_parameters.go:27]: function should return error if key inludes a quote char ["]

File: kusto/kql/query_parameters.go [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+		return ""
+	}
+
+	build := buildPool.Get().(*strings.Builder)
+	build.Reset()
+	defer buildPool.Put(build)
+
+	build.WriteString(declare)
+
+	for key, paramVals := range q.parameters {
```
Comment (yogilad) [inline kusto/kql/query_parameters.go:50]: x = false 
for 1 .. n 
    if x
       add ","
    x = true

File: kusto/kql/query_parameters.go [hunk 24]
```diff
@@ -0,0 +1,67 @@
+package kql
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+)
+
+type ParamVals struct {
```
Comment (yogilad) [inline kusto/kql/query_parameters.go:9]: This should be the value class

File: kusto/kql/strings.go [hunk 25]
```diff
@@ -0,0 +1,119 @@
+package kql
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/strings.go]: I don't like either of those names @yogilad, how is this related to string literals? these are string utilities/conversion functions

File: kusto/kql/query_parameters.go (hunk 21 above)
Comment (AsafMah) [reply kusto/kql/query_parameters.go:16]: I think it's a really good idea actually.

File: kusto/kql/query_parameters.go (hunk 23 above)
Comment (AsafMah) [reply kusto/kql/query_parameters.go:91]: I assume then it's putting the comma before the string?
@ronmonetaMicro you didn't implement this

File: kusto/kql/builder_test.go [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+			"MyTable | where i != \"foo\" ;",
+		},
+		{
+			"Test add string with quote",
+			NewStatementBuilder(
+				"MyTable | where i != ",
+			).AddString("foo\"bar").AddLiteral(" ;"),
+			"MyTable | where i != \"foo\\\"bar\" ;",
+		},
+		{"Test add identifiers",
```
Comment (AsafMah) [inline kusto/kql/builder_test.go:90]: Formatting

File: kusto/kql/query_parameters_test.go [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+			q := test.b.String()
+			params := test.qp.ToDeclarationString()
+			actual := fmt.Sprintf("%s\n%s", params, q)
+			for _, sub := range test.expected {
+				require.True(t, strings.Contains(actual, sub))
+			}
+		})
+	}
+}
+func TestToParameterCollection(t *testing.T) {
```
Comment (AsafMah) [inline kusto/kql/query_parameters_test.go:43]: Do one set of tests, for each check BOTH the declaration string and the parameter collection

File: kusto/kql/query_parameters_test.go [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+	dt, _ := time.Parse(time.RFC3339Nano, "2020-03-04T14:05:01.3109965Z")
```
Comment (AsafMah) [inline kusto/kql/query_parameters_test.go:44]: Still needs tests with unusual/complex values here

File: kusto/kql/query_parameters_test.go [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+
+func TestToDeclarationString(t *testing.T) {
+	tests := []struct {
+		name     string
+		b        Builder
+		qp       *StatementQueryParameters
+		failFlag bool
+		expected []string
+	}{
+		{"Test empty", NewStatementBuilder(""), NewStatementQueryParameters(), false, []string{"\n"}},
```
Comment (AsafMah) [inline kusto/kql/query_parameters_test.go:21]: Format it in new lines

File: kusto/kql/builder.go [hunk 30]
```diff
@@ -0,0 +1,115 @@
+package kql
```
Comment (AsafMah) [inline kusto/kql/builder.go:1]: Now that we are near the end it's time to update the docs with the new method:
- README.md
- sample (see stmt_example_test)
- doc.go
- update CHANGELOG.md (Make a ## [Unreleased] section)

File: kusto/kql/builder_test.go (hunk 26 above)
Comment (ronmonetaMicro) [reply kusto/kql/builder_test.go:90]: of which manner?

File: kusto/kql/query_parameters.go (hunk 23 above)
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters.go:91]: i've implemented it via lengths, instead of boolean flags + lengths

File: kusto/kql/builder_test.go (hunk 26 above)
Comment (AsafMah) [reply kusto/kql/builder_test.go:90]: New lines, look at the other tests in the file

File: kusto/kql/query_parameters.go (hunk 23 above)
Comment (AsafMah) [reply kusto/kql/query_parameters.go:91]: @yogilad WDYT?

File: kusto/doc.go [hunk 31]
```diff
@@ -56,6 +56,36 @@ and accepts only string constants for arguments.
...
 
+# Querying with parameters
+
+Users will sometimes want to use query parameters (Read more here - https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/queryparametersstatement?pivots=azuredataexplorer).
+In this case, the StatementBuilder and QueryParameters structs should be used.
+
+	dt, _ := time.Parse(time.RFC3339Nano, "2020-03-04T14:05:01.3109965Z")
+	// Query our database table "systemNodes" for the CollectionTimes and the NodeIds.
+	iter, err := client.Query(ctx, "database", kql.NewStatementBuilder("systemNodes | where CollectionTime == time and NodeId == id"),
+		[]kusto.QueryOption{kusto.QueryParameters(*kql.NewStatementQueryParameters().
```
Comment (AsafMah) [inline kusto/doc.go:67]: That's the wrong way to call the method, you don't need to create the array. Try it out and see

File: CHANGELOG.md [hunk 32]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+## [Unreleased]
+### Changed
```
Comment (AsafMah) [inline CHANGELOG.md:7]: Added, not changed

File: kusto/doc.go [hunk 33]
```diff
@@ -56,6 +56,36 @@ and accepts only string constants for arguments.
...
+							AddDateTime("time", dt).
+							AddInt("id", 1)))
+	if err != nil {
```
Comment (AsafMah) [inline kusto/doc.go:70]: Explanation for StatementBuilder?

File: kusto/kql/query_parameters.go (hunk 21 above)
Comment (AsafMah) [reply kusto/kql/query_parameters.go:16]: We should probably do it for StatementBuilder too, but probably out of scope here

**Review by AsafMah — COMMENTED** [review]
Good enough for me. 
Yochai has some open comments so I'll wait for him to go over & approve.

File: kusto/kql/query_parameters_test.go [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+			[]string{"declare",
+				"query_parameters(",
+				"databaseName:string",
+				"tableName:string",
+				"txt:string",
+				");\ndatabase(databaseName).table(tableName) | where column == txt ;"},
+			map[string]string{
+				"databaseName": "\"f\\\"\\\"o\"",
+				"tableName":    "\"b\\a\\r\"",
+				"txt":          "\"f_u_n\\u1234c\"",
```
Comment (yogilad) [inline kusto/kql/query_parameters_test.go:115]: nit: for the complex tests consider using Go String Literals quoted with a back-tick `this is a string literal`
You do not need to escape anything in such a string

https://www.digitalocean.com/community/tutorials/how-to-format-strings-in-go#raw-string-literals

File: kusto/kusto.go [hunk 35]
```diff
@@ -286,6 +280,17 @@ func setQueryOptions(ctx context.Context, op errors.Op, query Stmt, options ...Q
 			return nil, errors.ES(op, errors.KClientArgs, "QueryValues in the the Stmt were incorrect: %s", err).SetNoRetry()
 		}
 	}
+	if query.SupportsParameters() {
+		if len(opt.requestProperties.QueryParameters.ToParameterCollection()) != 0 {
```
Comment (yogilad) [inline kusto/kusto.go:284]: This needs to also be replaced with a simpler condition that does not require building the collection

File: kusto/kql/query_parameters.go (hunk 23 above)
Comment (yogilad) [reply kusto/kql/query_parameters.go:91]: Slightly more complex, but all in all ok

File: kusto/kql/query_parameters.go [hunk 36]
```diff
@@ -0,0 +1,109 @@
...
+import (
+	"github.com/Azure/azure-kusto-go/kusto/data/types"
+	"github.com/google/uuid"
+	"github.com/shopspring/decimal"
+	"strings"
+	"sync"
+	"time"
+)
+
+type StatementQueryParameters struct {
```
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters.go:12]: the name is indeed in use (and quite often...) so thought it's better to differentiate

[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/query_parameters.go:12]: I couldn't find any struct called QueryParameters, plus it's under a namespace.
I honestly think "kql.Parameters" is a good enough name, and if not then "kql.QueryParameters" is fine too.
@yogilad WDYT?

File: CHANGELOG.md [hunk 37]
```diff
@@ -3,6 +3,12 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+## [Unreleased]
+### Added
+* Created new Statement interface including both old Stmt and new statementBuilder struct for building queries
+* Added new StatementQueryParameters struct for simple and safe handling of query parameters
+### Changed
+* Old Stmt struct to adhere new Statement interface
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply CHANGELOG.md]: I'm not sure that we want to have a private struct name in there

File: README.md [hunk 38]
```diff
@@ -129,6 +129,62 @@ if err != nil {
...
+if err != nil {
+	panic("add error handling")
+}
+```
+#### Querying with StatementBuilder
+One way to build your query is using the StatementBuilder struct. With this struct you build your query as a combination of literals and special entities
+```go
+dt, _ := time.Parse(time.RFC3339Nano, "2020-03-04T14:05:01.3109965Z")
+// Query our database table "systemNodes" for the CollectionTimes and the NodeIds.
+iter, err := client.Query(ctx, "database", kql.NewStatementBuilder("").AddTable("systemNodes").AddLiteral(" | where ").AddColumn("CollectionTime").AddLiteral(" == ").AddDateTime(dt).AddLiteral(" and ").AddColumn("NodeId").AddLiteral(" == ").AddInt(1),
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply README.md]: Maybe add a comment for what the final query would be here?

Also, not yours, but iter.Do is deprecated, we should recommend using iter.DoOnRowOrError

[the remark this answers is not in the document]
Comment (AsafMah) [reply README.md]: In fact, we might not want to copy-paste the iter.Do on each one, it's just noise.

**Comment by AsafMah** [comment]
Another thing, since we don't have an interface behind StatementQueryParameters, we might want to be consistent and remove the Builder interface, and just have the statementBuilder struct be public and called Builder.

File: kusto/kql/query_parameters.go (hunk 21 above)
Comment (AsafMah) [reply kusto/kql/query_parameters.go:16]: @ronmonetaMicro after talking w/ Yochai, just remove the syncpool here too.

File: kusto/kql/strings.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/strings.go]: @ronmonetaMicro string_utils.go

File: README.md (hunk 38 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply README.md]: @ronmonetaMicro talked to Yohcai, just add the final query and that's it.

File: kusto/kql/query_parameters.go (hunk 36 above)
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/kql/query_parameters.go:12]: @ronmonetaMicro kql.Parameters

File: README.md (hunk 38 above)
[the remark this answers is not in the document]
Comment (ronmonetaMicro) [reply README.md]: not sure i follow - you mean like in the old version? 
iter, err := client.Query(ctx, "database", kql.NewStatementBuilder("").AddTable("systemNodes").AddLiteral(" | where ").AddColumn("CollectionTime").AddLiteral(" == ").AddDateTime(dt).AddLiteral(" and ").AddColumn("NodeId").AddLiteral(" == ").AddInt(1)

[the remark this answers is not in the document]
Comment (AsafMah) [reply README.md]: No, I mean add another line, showing that it has a .String() method that returns the full query.
Add in a comment what the result of it would be.

File: kusto/kql/query_parameters.go (hunk 21 above)
Comment (ronmonetaMicro) [reply kusto/kql/query_parameters.go:16]: not sure i understand, Whats the need in a reset? Strings already have a Reset function. and i didnt understand what change you wanted in Builder.go

File: kusto/kql/query_parameters.go [hunk 39]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// note - due to the psuedo-random nature of maps, the declaration string might be ordered differently for different runs.
+// might crash the test in those times.
+func (q *Parameters) ToDeclarationString() string {
+	const (
+		declare   = "declare query_parameters("
+		closeStmt = ");"
+	)
+	var build = new(strings.Builder)
```
Comment (AsafMah) [inline kusto/kql/query_parameters.go:77]: Use 	var build = strings.Builder{}
You don't need a pointer here

File: CHANGELOG.md [hunk 40]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+## [Unreleased]
+### Added
+* Added new builder struct for safe building of KQL statements from variables without use of 'Unsafe' mode.
```
Comment (AsafMah) [inline CHANGELOG.md:8]: Added kql.Builder

File: CHANGELOG.md [hunk 41]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
...
+* Simpler handling of query parameters using new parameters struct.
```
Comment (AsafMah) [inline CHANGELOG.md:9]: add kql.Parameters

File: README.md [hunk 42]
```diff
@@ -143,6 +143,69 @@ if err != nil {
...
+
+```go
+dt, _ := time.Parse(time.RFC3339Nano, "2020-03-04T14:05:01.3109965Z")
+// Query our database table "systemNodes" for the CollectionTimes and the NodeIds.
+statement := kql.NewStatementBuilder("systemNodes | where CollectionTime == time and NodeId == id")
+params :=  kql.NewStatementQueryParameters().AddDateTime("time", dt).AddInt("id", 1) 
+iter, err := client.Query(ctx, "database", statement, params)
+
+fmt.Println(fmt.Sprintf("%s\n%s", params.ToDeclarationString(), statement.String()))
+// Will print declare query_parameters(time:datetime, id:int)\nsystemNodes | where CollectionTime == time and NodeId == id
```
Comment (AsafMah) [inline README.md:159]: // Will print:
//  declare query_parameters(time:datetime, id:int);
//  systemNodes | where CollectionTime == time and NodeId == id
(In docs too)

**Review by AsafMah — COMMENTED** [review]
Approved, just fix small nit comments

File: kusto/conn.go [hunk 43]
```diff
@@ -155,7 +155,7 @@ func (c *Conn) doRequest(ctx context.Context, execType int, db string, query Sta
 	case execQuery, execMgmt:
 		var err error
 		var csl string
-		if query.SupportsParameters() || len(properties.QueryParameters.ToParameterCollection()) == 0 {
+		if query.SupportsInlineParameters() || properties.QueryParameters.Size() == 0 {
```
Comment (yogilad) [inline kusto/conn.go:158]: Count()

File: kusto/kql/query_parameters.go [hunk 44]
```diff
@@ -5,87 +5,81 @@ import (
...
-	New: func() interface{} {
-		return new(strings.Builder)
-	},
+func NewParameters() *Parameters {
+	return &Parameters{parameters: make(map[string]Value)}
 }
 
-func NewStatementQueryParameters() *StatementQueryParameters {
-	return &StatementQueryParameters{parameters: make(map[string]Value)}
+func (q *Parameters) Size() int {
```
Comment (yogilad) [inline kusto/kql/query_parameters.go:19]: Count()

**Review by yogilad — APPROVED** [review]
